File: src/commands/claim.ts
import { Command } from "commander";
import { requireAuth } from "../api.js";
import { success, error, info, color, formatDateTime } from "../utils/output.js";
import { randomUUID } from "crypto";

function parseRepoRef(ref: string): { store: string; repo: string; branch?: string } {
  const colonIdx = ref.indexOf(":");
  let repoRef = ref;
  let branch: string | undefined;

  if (colonIdx !== -1) {
    repoRef = ref.slice(0, colonIdx);
    branch = ref.slice(colonIdx + 1);
  }

  const parts = repoRef.split("/");
  if (parts.length !== 2) {
    throw new Error("Invalid reference. Use format: store/repo:branch");
  }

  return { store: parts[0], repo: parts[1], branch };
}

export function registerClaimCommand(program: Command): void {
  program
    .command("claim <store/repo:branch> <patterns...>")
    .description("Claim file patterns for editing (multi-agent coordination)")
    .option("-m, --message <message>", "Description of planned changes", "CLI claim")
    .option("--agent-id <id>", "Agent ID (auto-generated if not provided)")
    .option("--ttl <seconds>", "Claim TTL in seconds", "300")
    .addHelpText("after", `
Claim file patterns to prevent other agents from editing them.
Patterns use glob syntax (e.g., "src/**", "*.ts", "docs/api/*").

Format: store/repo:branch <patterns...>

Examples:
  scraps claim alice/my-project:main "src/**" -m "Refactoring"
  scraps claim alice/my-project:main "*.ts" "*.tsx" -m "Type fixes"
  scraps claim myteam/backend:main "api/**" --ttl 600 -m "API update"
  scraps claim alice/my-project:main "README.md" --agent-id my-agent-123

Options:
  -m, --message   Description shown to other agents (default: "CLI claim")
  --agent-id      Your agent ID (auto-generated if not provided, save for release)
  --ttl           Time-to-live in seconds before auto-release (default: 300)
`)
    .action(async (ref, patterns, opts) => {
      const client = requireAuth();
      const { store, repo, branch } = parseRepoRef(ref);

      if (!branch) {
        error("Branch is required. Use format: store/repo:branch");
        process.exit(1);
      }

      const agentId = opts.agentId || `cli-${randomUUID().slice(0, 8)}`;

      try {
        const result = await client.post(
          `/stores/${store}/repos/${repo}/branches/${encodeURIComponent(branch)}/coordinate/claim`,
          {
            agent_id: agentId,
            patterns,
            claim: opts.message,
            ttl_seconds: parseInt(opts.ttl),
          }
        );

        if (result.type === "claim_conflict") {
          error("Claim conflict detected:");
          for (const c of result.conflicts || []) {
            console.log(`  ${color(c.agent_name || c.agent_id?.slice(0, 8), "yellow")}: ${c.patterns?.join(", ")} - "${c.claim}"`);
          }
          process.exit(1);
        }

        success(`Claimed patterns: ${patterns.join(", ")}`);
        info(`Agent ID: ${agentId}`);
        if (result.expires_at) {
          info(`Expires: ${formatDateTime(new Date(result.expires_at))}`);
        }
      } catch (e: any) {
        error(`Failed to claim: ${e.message}`);
        process.exit(1);
      }
    });
}

File: internal/cli/claim.go
package cli

import (
	"fmt"

	"github.com/google/uuid"
	"github.com/spf13/cobra"

	"github.com/scraps-sh/scraps-cli/internal/api"
	"github.com/scraps-sh/scraps-cli/internal/config"
	"github.com/scraps-sh/scraps-cli/internal/model"
)

func newClaimCmd() *cobra.Command {
	var message, agentID string
	var ttl int

	cmd := &cobra.Command{
		Use:   "claim <store/repo:branch> <patterns...>",
		Short: "Claim file patterns for exclusive access",
		Args:  cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			store, repo, branch, err := parseStoreRepoBranch(args[0])
			if err != nil {
				return err
			}

			if branch == "" {
				return fmt.Errorf("branch is required (use store/repo:branch format)")
			}

			patterns := args[1:]

			// Generate agent ID if not provided
			if agentID == "" {
				agentID = "cli-" + uuid.New().String()[:8]
			}

			if message == "" {
				message = "CLI claim"
			}

			client, err := api.NewClientFromConfig("")
			if err != nil {
				return err
			}

			req := model.ClaimRequest{
				AgentID:    agentID,
				Patterns:   patterns,
				Claim:      message,
				TTLSeconds: ttl,
			}

			resp, err := client.Claim(store, repo, branch, req)
			if err != nil {
				return err
			}

			// Check for conflicts
			if resp.Type == "claim_conflict" && len(resp.Conflicts) > 0 {
				errorf("Claim conflict detected!")
				fmt.Println("\nConflicting claims:")
				for _, c := range resp.Conflicts {
					fmt.Printf("  Agent: %s (%s)\n", c.AgentName, c.AgentID)
					fmt.Printf("  Patterns: %v\n", c.Patterns)
					fmt.Printf("  Claim: %s\n\n", c.Claim)
				}
				return fmt.Errorf("cannot claim: patterns conflict with existing claims")
			}

			if config.GetOutputFormat() == "json" {
				outputJSON(map[string]any{
					"agent_id":   agentID,
					"patterns":   patterns,
					"expires_at": resp.ExpiresAt,
				})
			} else {
				success(fmt.Sprintf("Claimed patterns as %s", agentID))
				fmt.Printf("Patterns: %v\n", patterns)
				if resp.ExpiresAt != nil {
					fmt.Printf("Expires: %s\n", *resp.ExpiresAt)
				}
				info(fmt.Sprintf("Use --agent-id %s to release", agentID))
			}

			return nil
		},
	}

	cmd.Flags().StringVarP(&message, "message", "m", "", "Claim description")
	cmd.Flags().StringVar(&agentID, "agent-id", "", "Agent ID (auto-generated if not provided)")
	cmd.Flags().IntVar(&ttl, "ttl", 300, "Claim TTL in seconds")

	return cmd
}

func newReleaseCmd() *cobra.Command {
	var agentID string

	cmd := &cobra.Command{
		Use:   "release <store/repo:branch> <patterns...>",
		Short: "Release claimed file patterns",
		Args:  cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			store, repo, branch, err := parseStoreRepoBranch(args[0])
			if err != nil {
				return err
			}

			if branch == "" {
				return fmt.Errorf("branch is required (use store/repo:branch format)")
			}

			if agentID == "" {
				return fmt.Errorf("--agent-id is required")
			}

			patterns := args[1:]

			client, err := api.NewClientFromConfig("")
			if err != nil {
				return err
			}

			req := model.ReleaseRequest{
				AgentID:  agentID,
				Patterns: patterns,
			}

			if err := client.Release(store, repo, branch, req); err != nil {
				return err
			}

			success(fmt.Sprintf("Released patterns as %s", agentID))
			return nil
		},
	}

	cmd.Flags().StringVar(&agentID, "agent-id", "", "Agent ID from the claim")
	cmd.MarkFlagRequired("agent-id")

	return cmd
}
